src/main/kotlin/flag/FlagConfigStreamApi.kt

```diff
@@ -0,0 +1,146 @@
+package com.amplitude.experiment.flag
+
+import com.amplitude.experiment.evaluation.EvaluationFlag
+import com.amplitude.experiment.util.*
+import com.amplitude.experiment.util.SseStream
+import kotlinx.serialization.decodeFromString
+import okhttp3.HttpUrl
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.sse.EventSource
+import okhttp3.sse.EventSourceListener
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.ExecutionException
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
+
+internal open class FlagConfigStreamApiError(message: String?, cause: Throwable?): Exception(message, cause) {
+    constructor(message: String?) : this(message, null)
+    constructor(cause: Throwable?) : this(cause?.toString(), cause)
+}
+internal class FlagConfigStreamApiConnTimeoutError: FlagConfigStreamApiError("Initial connection timed out")
+internal class FlagConfigStreamApiDataCorruptError: FlagConfigStreamApiError("Stream data corrupted")
+internal class FlagConfigStreamApiStreamError(cause: Throwable?): FlagConfigStreamApiError("Stream error", cause)
+
+private const val CONNECTION_TIMEOUT_MILLIS_DEFAULT = 1500L
+private const val KEEP_ALIVE_TIMEOUT_MILLIS_DEFAULT = 17000L // keep alive sends at 15s interval. 2s grace period
+private const val RECONN_INTERVAL_MILLIS_DEFAULT = 15 * 60 * 1000L
+internal class FlagConfigStreamApi (
+    deploymentKey: String,
+    serverUrl: HttpUrl,
+    httpClient: OkHttpClient = OkHttpClient(),
+    val connectionTimeoutMillis: Long = CONNECTION_TIMEOUT_MILLIS_DEFAULT,
+    keepaliveTimeoutMillis: Long = KEEP_ALIVE_TIMEOUT_MILLIS_DEFAULT,
+    reconnIntervalMillis: Long = RECONN_INTERVAL_MILLIS_DEFAULT,
+) {
+    private val lock: ReentrantLock = ReentrantLock()
+    var onInitUpdate: ((List<EvaluationFlag>) -> Unit)? = null
+    var onUpdate: ((List<EvaluationFlag>) -> Unit)? = null
+    var onError: ((Exception?) -> Unit)? = null
+    val url = serverUrl.newBuilder().addPathSegments("sdk/stream/v1/flags").build()
+    private val stream: SseStream = SseStream(
+        "Api-Key $deploymentKey",
+        url,
+        httpClient,
+        connectionTimeoutMillis,
+        keepaliveTimeoutMillis,
+        reconnIntervalMillis)
+
+    internal fun connect() {
+        // Guarded by lock. Update to callbacks and waits can lead to race conditions.
+        lock.withLock {
+            val isInit = AtomicBoolean(true)
+            val connectTimeoutFuture = CompletableFuture<Unit>()
+            val updateTimeoutFuture = CompletableFuture<Unit>()
+            stream.onUpdate = { data ->
+                if (isInit.getAndSet(false)) {
```

Comment:
This boolean logic feels like the opposite of what I would expect. Why would receiving an update set the `isInit` boolean to false? wouldn't an update mean that the stream is initialized?

Reply:
It means if it's during an initialization process. Receiving an update terminates initialization. Renamed to `isDuringInit`.